cmd/metal-api/internal/service/machine-service.go

```diff
@@ -1555,6 +1578,129 @@ func freeMachine(ds *datastore.RethinkStore, publisher bus.Publisher, ipamer ipa
 	return nil
 }
 
+// reinstallMachine reinstalls the requested machine with given image by either allocating
+// the machine if not yet allocated or not modifying any other allocation parameter than 'ImageID'
+// and 'Reinstall' set to true.
+// If the given image ID is nil, it deletes the machine instead.
+func (r machineResource) reinstallMachine(request *restful.Request, response *restful.Response) {
+	var requestPayload v1.MachineReinstallRequest
+	err := request.ReadEntity(&requestPayload)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	id := request.PathParameter("id")
+	m, err := r.ds.FindMachineByID(id)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	logger := utils.Logger(request).Sugar()
+
+	var resp *v1.MachineResponse
+	if m.Allocation != nil {
+		old := *m
+
+		m.Allocation.Reinstall = true
+		m.Allocation.ImageID = requestPayload.ImageID
+
+		resp = makeMachineResponse(m, r.ds, logger)
+		if resp.Allocation.Image != nil {
+			err = r.ds.UpdateMachine(&old, m)
+			if checkError(request, response, utils.CurrentFuncName(), err) {
+				return
+			}
+			logger.Infow("marked machine to get reinstalled", "machineID", m.ID)
+
+			err = deleteVRFSwitches(r.ds, m, logger)
+			if checkError(request, response, utils.CurrentFuncName(), err) {
+				return
+			}
+
+			err = publishDeleteEvent(r, m, logger)
+			if checkError(request, response, utils.CurrentFuncName(), err) {
+				return
+			}
+
+			err = publishMachineCmd(logger, m, r, metal.MachineReinstall)
+			if err != nil {
+				logger.Errorw("unable to publish machine command", "command", metal.MachineReinstall, "machineID", m.ID, "error", err)
+			}
+		}
+	} else {
+		resp = makeMachineResponse(m, r.ds, logger)
```

Comment:
So... you can call this endpoint and when there is no allocation just nothing happens and the result is "ok"? Wouldn't it be better to return an error if there is no allocation?